keras_nlp/utils/text_generation.py

```diff
@@ -128,30 +132,69 @@ def token_probability_fn(inputs):
     ```
 
     """
-    if not tf.executing_eagerly():
-        raise RuntimeError(
-            "`keras_nlp.utils.greedy_search` currently requires an eager "
-            "execution context. Please call `greedy_search` outside "
-            "tf.function or run `tf.config.run_functions_eagerly(True)` to run "
-            "tf.function in eager mode."
-        )
-
     prompt = validate_prompt(prompt)
 
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
     validate_token_probability_fn(token_probability_fn, prompt)
 
-    i = prompt.shape[1]
-    while i < max_length:
-        # If the prompt has reached our desired length, exit while loop.
-        pred = token_probability_fn(prompt)
+    initial_shape = prompt.shape.as_list()
+    initial_length = initial_shape[1]
+
+    # Pad the prompt with `pad_token_id` to `max_length`. We use `map_fn` here
+    # because the batch_size might not be static.
+    prompt = tf.map_fn(
+        fn=lambda sequence: tf.concat(
+            [sequence, tf.fill([max_length - initial_length], pad_token_id)],
+            axis=0,
+        ),
+        elems=prompt,
+        fn_output_signature=tf.TensorSpec(
+            shape=(max_length), dtype=prompt.dtype
+        ),
+    )
+
+    # Create a state dictionary.
+    state = {"length": initial_length, "prompt": prompt}
+
+    def one_step(state):
+        length = state["length"]
+        prompt = state["prompt"]
+
+        pred = token_probability_fn(prompt[:, :length])
         next_token = tf.cast(tf.argmax(pred, axis=-1), dtype=prompt.dtype)
+
         # Append the next token to current sequence.
-        prompt = tf.concat([prompt, next_token[:, tf.newaxis]], axis=-1)
-        i += 1
+        def add_token(args):
+            sequence, token = args
+            return tf.tensor_scatter_nd_update(
+                tensor=sequence, indices=[[length]], updates=[token]
+            )
+
+        prompt = tf.map_fn(
+            fn=add_token,
+            elems=(prompt, next_token),
+            fn_output_signature=tf.TensorSpec(
+                shape=(max_length), dtype=prompt.dtype
+            ),
+        )
 
+        state["length"] += 1
+        state["prompt"] = prompt
+        return [state]
+
+    # Run a while loop till text of length `max_length` has been generated.
+    state = tf.while_loop(
+        cond=lambda state: tf.less(
+            tf.cast(state["length"], dtype=tf.int64),
+            tf.cast(max_length, dtype=tf.int64),
+        ),
+        body=one_step,
+        loop_vars=[state],
+    )[0]
+
+    prompt = state["prompt"]
     if end_token_id is not None:
         prompt = mask_tokens_after_end_token(
```

Comment:
do we even need this function anymore, if we are just starting with the correct sized tensor filled with pad_token_id?

Reply:
hmm I guess we do to avoid random tokens after the end_token_id

Reply:
Yep